### PR TITLE
[FLINK-11334][core] Migrate enum serializers to use new serialization compatibility abstractions

### PR DESCRIPTION
**Pull request by aljoscha**
This is an updated version of #7710.

It has the following changes on top:
 - Remove old deserialization logic from new EnumSerializer Snapshot
 - Fix compatibility check in EnumSerializer Snapshot and do proper reconfiguration

The individual commits have explanations for why the change is necessary.

cc @klion26

### CODE REVIEW

**Comment by klion26** [comment]
Thank you @aljoscha, the code looks good to me.

File: flink-core/src/test/java/org/apache/flink/api/common/typeutils/TypeSerializerSnapshotMigrationTestBase.java [hunk 1]
```diff
@@ -452,11 +490,12 @@ protected boolean matchesSafely(TypeSerializerSchemaCompatibility<T> testResultC
 				if (expectedCompatibilty.isCompatibleAsIs()) {
 					return testResultCompatibility.isCompatibleAsIs();
 				} else if (expectedCompatibilty.isIncompatible()) {
-					return testResultCompatibility.isCompatibleAfterMigration();
-				} else if (expectedCompatibilty.isIncompatible()) {
-					return testResultCompatibility.isIncompatible();
+					return testResultCompatibility.isCompatibleAfterMigration() ||
+							testResultCompatibility.isIncompatible();
```
Comment (tzulitai) [inline flink-core/src/test/java/org/apache/flink/api/common/typeutils/TypeSerializerSnapshotMigrationTestBase.java]: looks a bit odd.

If think it's supposed to be
```
...
} else if (expectedCompatibilty.isIncompatible()) {
    return testResultCompatibility.isIncompatible();
} else if (expectedCompatibilty.isCompatibleAfterMigration()) {
    return testResultCompatibility.isCompatibleAfterMigration();
}
...
```

Comment (aljoscha) [reply flink-core/src/test/java/org/apache/flink/api/common/typeutils/TypeSerializerSnapshotMigrationTestBase.java]: yes, I think I merged this version in the end 😅 (I mean the correct one)

**Comment by aljoscha** [comment]
Merged

**Pull request closed without merging**